### PR TITLE
Content migration batch 6

### PR DESCRIPTION
**Pull request by mattletford**
Content migration

From: https://developers.google.com/web/fundamentals/push-notifications/common-notification-patterns
To: https://web.dev/push-notifications-common-notification-patterns

From: https://developers.google.com/web/fundamentals/push-notifications/display-a-notification
To: https://web.dev/push-notifications-display-a-notification

From: https://developers.google.com/web/fundamentals/push-notifications/how-push-works
To: https://web.dev/push-notifications-how-push-works

From: https://developers.google.com/web/fundamentals/push-notifications/notification-behaviour
To: https://web.dev/push-notifications-notification-behaviour

From: https://developers.google.com/web/fundamentals/native-hardware/user-location
To: https://web.dev/push-notifications-common-issues-and-reporting-bugs

@rachelandrew

### CODE REVIEW

File: src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+[Raising Bug Reports](#raising_bug_reports)
+section to file a good bug report with all the necessary information to expedite
+the bug fixing process.
+
+One thing I'd like to call out before we start is that **Firefox and the
+Mozilla AutoPush Service have great error messages.** If you get stuck and
+are not sure what the problem is, then test in Firefox and see if you
+get a more helpful error message.
+
+## Authorization Issues
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md:53]:
```suggestion
## Authorization issues
```

File: src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+starting out with web push. This is normally a problem with the configuration of a
+site's [Application Server Keys (a.k.a. VAPID keys)
+](https://tools.ietf.org/html/draft-ietf-webpush-vapid-02).
+
+The easiest way to support push in both Firefox and Chrome is to supply an
+`applicationServerKey` in the `subscribe()` call. The down side is that
+any discrepancy between your front end and server's keys will result in an
+authorization error.
+
+### On Chrome + FCM
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md:65]:
```suggestion
### On Chrome and FCM
```

File: src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+{
+  "code": 401,
+  "errno": 109,
+  "error": "Unauthorized",
+  "more_info": "http://autopush.readthedocs.io/en/latest/http.html#error-codes",
+  "message": "Request did not validate Invalid Authorization Header"
+}
+```
+
+## HTTP Status Codes
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md:160]:
```suggestion
## HTTP status codes
```

File: src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+</tr>
+</table>
+
+If the HTTP status code is not in this list and the error message is not
+helpful, check the [Web Push Protocol
+spec](https://tools.ietf.org/html/draft-ietf-webpush-protocol) to see if the
+status code is referenced along with a scenario of when that status code can
+be used.
+
+## Payload Encryption Issue
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md:206]:
```suggestion
## Payload encryption issue
```

File: src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+If you can successfully trigger a push message (i.e. send a message to a web
+push service and receive a 201 response code) but the push event never fires in
+your service worker, this normally indicates that the browser failed to
+decrypt the message it received.
+
+If this is the case, you should see an error message in Firefox's DevTools
+console like so:
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/g8VpxsP7PMWPSdtSvAkY.png", alt="Firefox DevTools with decryption message", width="800", height="269" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md:216]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/g8VpxsP7PMWPSdtSvAkY.png", alt="Firefox DevTools with decryption message.", width="800", height="269" %}
```

File: src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+To check if this is the issue in Chrome, do the following:
+
+1. Go to about://gcm-internals and click the "Start Recording" button.
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/VH4juz336de2yiInKjUO.png", alt="Chrome GCM internals record", width="800", height="226" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md:222]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/VH4juz336de2yiInKjUO.png", alt="Chrome GCM internals record.", width="800", height="226" %}
```

File: src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+
+1. Trigger a push message, and look under the "Message Decryption Failure Log".
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/DszP9aaiMTmX5LdTXKPx.png", alt="GCM internals decryption log", width="800", height="155" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md:226]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/DszP9aaiMTmX5LdTXKPx.png", alt="GCM internals decryption log.", width="800", height="155" %}
```

File: src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+If there is an issue with the decryption of the payload, you'll see an error
+similar to the one displayed above. (Notice the `AES-GCM decryption failed`
+message in the details column.)
+
+There are a few tools which may help debug encryption if this is your issue:
+
+- [Push Encryption Verifier tool by Peter Beverloo](https://tests.peter.sh/push-encryption-verifier/).
+- [Web Push: Data Encryption Test Page by Mozilla](https://mozilla-services.github.io/WebPushDataTestPage/)
+
+## Connection Issue
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md:237]:
```suggestion
## Connection issue
```

File: src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+
+If you aren't receiving a push event in your service worker and you aren't
+seeing any decryption errors, then the browser may be failing to connect to
+a push service.
+
+In Chrome, you can check whether the browser is receiving messages by examining
+the 'Receive Message Log' (sic) in `about://gcm-internals`.
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/vRKzAO05KD5vUpQ6nL3z.png", alt="GCM internals receive message log", width="800", height="117" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md:246]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/vRKzAO05KD5vUpQ6nL3z.png", alt="GCM internals receive message log.", width="800", height="117" %}
```

File: src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+
+If you aren't seeing the message arrive in a timely fashion, then make sure that
+the connection status of your browser is `CONNECTED`:
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/YdYj19Kn0zl6Jlfbm9hg.png", alt="GCM internals connection state", width="584", height="246" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md:251]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/YdYj19Kn0zl6Jlfbm9hg.png", alt="GCM internals connection state.", width="584", height="246" %}
```

File: src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+
+If it's **not** 'CONNECTED', you may need to delete your current profile and
+[create a new one](https://support.google.com/chrome/answer/2364824). If that
+still doesn't solve the issue, please raise a bug report as suggested below.
+
+## Raising Bug Reports
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-common-issues-and-reporting-bugs/index.md:257]:
```suggestion
## Raising bug reports
```

File: src/site/content/en/blog/push-notifications-common-notification-patterns/index.md [hunk 12]
```diff
@@ -0,0 +1,402 @@
+---
+title: Common Notification Patterns
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-common-notification-patterns/index.md:2]:
```suggestion
title: Common notification patterns
```

File: src/site/content/en/blog/push-notifications-common-notification-patterns/index.md [hunk 13]
```diff
@@ -0,0 +1,402 @@
...
+  if (matchingClient) {
+    return matchingClient.focus();
+  } else {
+    return clients.openWindow(urlToOpen);
+  }
+});
+```
+
+{% Aside %}
+Note: We are returning the promise for `matchingClient.focus()` and
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-common-notification-patterns/index.md:190]:
```suggestion
We are returning the promise for `matchingClient.focus()` and
```

File: src/site/content/en/blog/push-notifications-common-notification-patterns/index.md [hunk 14]
```diff
@@ -0,0 +1,402 @@
...
+});
+```
+
+If there is a notification currently displayed, we increment the message count and set the
+notification title and body message accordingly. If there
+are no notifications, we create a new notification with a `newMessageCount` of 1.
+
+The result is that the first message would look like this:
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/plgFelmhahN06Unqr2dr.png", alt="First notification without merging", width="380", height="116" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-common-notification-patterns/index.md:278]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/plgFelmhahN06Unqr2dr.png", alt="First notification without merging.", width="380", height="116" %}
```

File: src/site/content/en/blog/push-notifications-common-notification-patterns/index.md [hunk 15]
```diff
@@ -0,0 +1,402 @@
...
+
+A second notification would collapse the notifications into this:
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/bKoZiyLJI8DVYOPQBIGv.png", alt="Second notification with merging", width="380", height="121" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-common-notification-patterns/index.md:282]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/bKoZiyLJI8DVYOPQBIGv.png", alt="Second notification with merging.", width="380", height="121" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 16]
```diff
@@ -0,0 +1,414 @@
...
+to test them out for yourself then click the button below.
+
+<a class="button" href="https://web-push-book.gauntface.com/demos/notification-examples/" target="_blank">Notification Demos</a>
+
+## Visual Options
+
+The API for showing a notification is simply:
+
+```xml
+    <ServiceWorkerRegistration>.showNotification(<title>, <options>);
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:26]:
```suggestion
<ServiceWorkerRegistration>.showNotification(<title>, <options>);
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 17]
```diff
@@ -0,0 +1,414 @@
...
+  "actions": "<Array of Strings>",
+
+  "//": "Information Option. No visual affect.",
+  "timestamp": "<Long>"
+}
+```
+
+First let's look at the visual options.
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/ZVIQcUqIlNerdPiqOM9C.png", alt="Dissection of the UI of a Notification", width="800", height="284" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:59]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/ZVIQcUqIlNerdPiqOM9C.png", alt="Dissection of the UI of a Notification.", width="800", height="284" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 18]
```diff
@@ -0,0 +1,414 @@
...
+
+### Title and Body Options
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:61]:
```suggestion
### Title and body options
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 19]
```diff
@@ -0,0 +1,414 @@
...
+const title = 'Simple Title';
+const options = {
+  body: 'Simple piece of body text.\nSecond line of body text :)',
+};
+registration.showNotification(title, options);
+```
+
+We'd get this notification on Chrome:
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/BZnd4sFC3XMq8yPUkYVz.png", alt="Notification with title and body text on Chrome on Linux", width="380", height="117" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:78]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/BZnd4sFC3XMq8yPUkYVz.png", alt="Notification with title and body text on Chrome on Linux.", width="380", height="117" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 20]
```diff
@@ -0,0 +1,414 @@
...
+
+On Firefox on Linux it would look like this:
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/hyJWXo3xO7AlrlE3I61F.png", alt="Notification with title and body text on Firefox on Linux", width="521", height="74" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:82]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/hyJWXo3xO7AlrlE3I61F.png", alt="Notification with title and body text on Firefox on Linux.", width="521", height="74" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 21]
```diff
@@ -0,0 +1,414 @@
...
+
+I was curious about what would happen if I added lots of text and this was the result:
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/MjZfnuD9h6MJLjv9enVi.png", alt="Notification with long title and body text on Chrome on Linux", width="380", height="156" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:86]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/MjZfnuD9h6MJLjv9enVi.png", alt="Notification with long title and body text on Chrome on Linux.", width="380", height="156" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 22]
```diff
@@ -0,0 +1,414 @@
...
+
+Interestingly, Firefox on Linux collapses the body text until you hover the notification,
+causing the notification to expand.
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/hyJWXo3xO7AlrlE3I61F.png", alt="Notification with long title and body text on Firefox on Linux", width="521", height="74" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:91]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/hyJWXo3xO7AlrlE3I61F.png", alt="Notification with long title and body text on Firefox on Linux.", width="521", height="74" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 23]
```diff
@@ -0,0 +1,414 @@
...
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/Huyo4xr85SwcmPN9zSly.png", alt="Notification with long title and body text on Firefox on Linux while hovering over the notification with the mouse cursor", width="521", height="159" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:93]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/Huyo4xr85SwcmPN9zSly.png", alt="Notification with long title and body text on Firefox on Linux while hovering over the notification with the mouse cursor.", width="521", height="159" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 24]
```diff
@@ -0,0 +1,414 @@
...
+The reason I've included these examples is twofold. There will be differences
+between browsers. Just looking at text, Firefox and Chrome look and act
+differently. Secondly there are differences across platforms. Chrome has a
+custom UI for all platforms whereas Firefox uses the system notifications on
+my Linux machine. The same notifications on Windows with Firefox look
+like this:
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/OQRchMO68ph6UV2Lovy0.png", alt="Notification with title and body text on Firefox on Windows", width="562", height="281" %}
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/OQRchMO68ph6UV2Lovy0.png", alt="Notification with long title and body text on Firefox on Windows", width="562", height="281" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:104]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/OQRchMO68ph6UV2Lovy0.png", alt="Notification with title and body text on Firefox on Windows.", width="562", height="281" %}

{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/OQRchMO68ph6UV2Lovy0.png", alt="Notification with long title and body text on Firefox on Windows.", width="562", height="281" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 25]
```diff
@@ -0,0 +1,414 @@
...
+const title = 'Icon Notification';
+const options = {
+  icon: '/images/demos/icon-512x512.png',
+};
+registration.showNotification(title, options);
+```
+
+On Chrome we get this notification on Linux:
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/RBQH1QGLIRT9HxJTDjkj.png", alt="Notification with icon on Chrome on Linux", width="380", height="100" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:122]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/RBQH1QGLIRT9HxJTDjkj.png", alt="Notification with icon on Chrome on Linux.", width="380", height="100" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 26]
```diff
@@ -0,0 +1,414 @@
...
+
+and on Firefox:
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/EoS5ydulQRjq1nxTJxrA.png", alt="Notification with icon on Firefox on Linux", width="522", height="75" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:126]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/EoS5ydulQRjq1nxTJxrA.png", alt="Notification with icon on Firefox on Linux.", width="522", height="75" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 27]
```diff
@@ -0,0 +1,414 @@
...
+Sadly there aren't any solid guidelines for what size image to use for an icon.
+
+[Android seems to want a 64dp image](http://stackoverflow.com/questions/7220738/honeycomb-notifications-how-to-set-largeicon-to-the-right-size)
+(which is 64px multiples by the device pixel ratio).
+
+If we assume the highest pixel ratio for a device will be 3, an icon size
+of 192px or more is a safe bet.
+
+{% Aside %}
+Note: Some browsers may require the image be served over HTTPS. Be aware of this
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:137]:
```suggestion
Some browsers may require the image be served over HTTPS. Be aware of this
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 28]
```diff
@@ -0,0 +1,414 @@
...
+const title = 'Badge Notification';
+const options = {
+  badge: '/images/demos/badge-128x128.png',
+};
+registration.showNotification(title, options);
+```
+
+At the time of writing the badge is only used on Chrome for Android.
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/LaAYwycbWMn8SCro7fyo.png", alt="Notification with badge on Chrome for Android", width="800", height="248" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:156]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/LaAYwycbWMn8SCro7fyo.png", alt="Notification with badge on Chrome for Android.", width="800", height="248" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 29]
```diff
@@ -0,0 +1,414 @@
...
+const title = 'Image Notification';
+const options = {
+  image: '/images/demos/unsplash-farzad-nazifi-1600x1100.jpg',
+};
+registration.showNotification(title, options);
+```
+
+On desktop the notification will look like this:
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/VF32ZZXDZj2xO7dJdR3W.png", alt="Notification with image on Chrome on Linux", width="800", height="554" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:185]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/VF32ZZXDZj2xO7dJdR3W.png", alt="Notification with image on Chrome on Linux.", width="800", height="554" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 30]
```diff
@@ -0,0 +1,414 @@
...
+
+On Android the cropping and ratio are different.
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/xF6oPjieHKCqtvrsWfM7.png", alt="Notification with image on Chrome for Android", width="380", height="340" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:189]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/xF6oPjieHKCqtvrsWfM7.png", alt="Notification with image on Chrome for Android.", width="380", height="340" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 31]
```diff
@@ -0,0 +1,414 @@
...
+  options.body =
+    `This notification will only display ` + `${maxVisibleActions} actions.`;
+} else {
+  options.body =
+    `This notification can display up to ` + `${maxVisibleActions} actions.`;
+}
+```
+
+registration.showNotification(title, options);
+
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:246]:
```suggestion
}

registration.showNotification(title, options);
```

```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 32]
```diff
@@ -0,0 +1,414 @@
...
+At the time of writing only Chrome and Opera for Android support actions.
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/GI8sj9krxVtxWVeHGuvs.png", alt="Notification with actions on Chrome for Android", width="800", height="296" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:249]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/GI8sj9krxVtxWVeHGuvs.png", alt="Notification with actions on Chrome for Android.", width="800", height="296" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 33]
```diff
@@ -0,0 +1,414 @@
...
+
+In the example above I've defined 4 actions to illustrate that you can define more actions than
+will be displayed. If you want to know the number actions that will be displayed by the browser,
+you can check `Notification.maxActions`, which is used in the body text in the demo.
+
+On desktop, the action button icons display their colors (see the pink doughnut above).
+
+On Android Marshmallow, the icons are colored to match the system color scheme:
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/REyUtJ7icGvOoepXgpLQ.png", alt="Notification with actions on Chrome on Linux", width="380", height="178" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:263]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/REyUtJ7icGvOoepXgpLQ.png", alt="Notification with actions on Chrome on Linux.", width="380", height="178" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 34]
```diff
@@ -0,0 +1,414 @@
...
+to layout options like actions, but I saw no difference.
+
+Probably best to define if you can, otherwise the browser should do the right thing according
+to the text supplied.
+
+The parameter should be set to either `auto`, `ltr` or `rtl`.
+
+A right-to-left language used on Chrome on Linux looks like this:
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/epWm0UDjQzcRMz37uaaa.png", alt="Notification with right-to-left language on Chrome on Linux", width="380", height="234" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:304]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/epWm0UDjQzcRMz37uaaa.png", alt="Notification with right-to-left language on Chrome on Linux.", width="380", height="234" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 35]
```diff
@@ -0,0 +1,414 @@
...
+
+On Firefox (while hovering over it) you'll get this:
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/ncKq2yMj7ipFkGGOWKBr.png", alt="Notification with right-to-left language on Firefox on Linux", width="521", height="102" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:308]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/ncKq2yMj7ipFkGGOWKBr.png", alt="Notification with right-to-left language on Firefox on Linux.", width="521", height="102" %}
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 36]
```diff
@@ -0,0 +1,414 @@
...
+
+### Vibrate
+
+{% Aside 'caution' %}
+Caution: Support for [`vibrate` has been deprecated](https://crbug.com/971422)
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:313]:
```suggestion
Support for [`vibrate` has been deprecated](https://crbug.com/971422)
```

File: src/site/content/en/blog/push-notifications-display-a-notification/index.md [hunk 37]
```diff
@@ -0,0 +1,414 @@
...
+Some common pitfalls to avoid:
+
+- Don't put your website in the title or the body. Browsers include your domain in the
+  notification so **don't duplicate it**.
+- Use all the information that you have available. If you send a push message because someone
+  sent a message to a user, rather than using a title of 'New Message' and body of 'Click here to
+  read it.' use a title of 'John just sent a new message' and set the body of the notification to
+  part of the message.
+
+## Browsers and Feature Detection
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-display-a-notification/index.md:391]:
```suggestion
## Browsers and feature detection
```

File: src/site/content/en/blog/push-notifications-how-push-works/index.md [hunk 38]
```diff
@@ -0,0 +1,154 @@
+---
+title: How Push Works
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-how-push-works/index.md:2]:
```suggestion
title: How push works
```

File: src/site/content/en/blog/push-notifications-how-push-works/index.md [hunk 39]
```diff
@@ -0,0 +1,154 @@
...
+push service to know which application server subscribed a user and ensure that it's the same
+server triggering the push messages to that user.
+
+Once you've subscribed the user and have a `PushSubscription`, you'll need to send the
+`PushSubscription` details to your backend / server. On your server, you'll save this
+subscription to a database and use it to send a push message to that user.
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/3eZW2SvRzowpHsYxhUwr.svg", alt="Make sure you send the PushSubscription to your backend.", width="800", height="213" %}
+
+## Step 2: Send a Push Message
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-how-push-works/index.md:48]:
```suggestion
## Step 2: Send a push message
```

File: src/site/content/en/blog/push-notifications-how-push-works/index.md [hunk 40]
```diff
@@ -0,0 +1,154 @@
...
+
+Some questions you might be asking yourself:
+
+- Who and what is the push service?
+
+- What does the API look like? Is it JSON, XML, something else?
+
+- What can the API do?
+
+### Who and What is the Push Service?
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-how-push-works/index.md:62]:
```suggestion
- Who and what is the push service?
- What does the API look like? Is it JSON, XML, something else?
- What can the API do?

### Who and what is the push service?
```

File: src/site/content/en/blog/push-notifications-how-push-works/index.md [hunk 41]
```diff
@@ -0,0 +1,154 @@
...
+
+- The time-to-live for a push message. This defines how long a message should be queued before
+  it's removed and not delivered.
+
+- Define the urgency of the message. This is useful in case the push service is preserving the
+  users battery life by only delivering high priority messages.
+
+- Give a push message a "topic" name which will replace any pending message with this new message.
+
+{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/f42EMdaS3bjYhIw47E1J.svg", alt="When your server wishes to send a push message, it makes a web push protocol request to a push service", width="800", height="220" %}
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-how-push-works/index.md:127]:
```suggestion
{% Img src="image/C47gYyWYVMMhDmtYSLOWazuyePF2/f42EMdaS3bjYhIw47E1J.svg", alt="When your server wishes to send a push message, it makes a web push protocol request to a push service.", width="800", height="220" %}
```

File: src/site/content/en/blog/push-notifications-how-push-works/index.md [hunk 42]
```diff
@@ -0,0 +1,154 @@
...
+
+### Step 3: Push Event on the User's Device
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-how-push-works/index.md:129]:
```suggestion
### Step 3: Push event on the user's device
```

File: src/site/content/en/blog/push-notifications-notification-behaviour/index.md [hunk 43]
```diff
@@ -0,0 +1,270 @@
+---
+title: Notification Behavior
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-notification-behaviour/index.md:2]:
```suggestion
title: Notification behavior
```

File: src/site/content/en/blog/push-notifications-notification-behaviour/index.md [hunk 44]
```diff
@@ -0,0 +1,270 @@
...
+const title = 'Silent Notification';
+const options = {
+  silent: true,
+};
+registration.showNotification(title, options);
+```
+
+Note: If you define both `silent` and `renotify`, `silent` will take precedence.
+
+## Requires Interaction
```
Comment (rachelandrew) [inline src/site/content/en/blog/push-notifications-notification-behaviour/index.md:245]:
```suggestion
## Requires interaction
```

**Review by rachelandrew — CHANGES_REQUESTED** [review]
You should be able to go through and just commit these as a batch.

**Comment by rachelandrew** [comment]
@mattletford please follow the process detailed here to add suggestions to a batch https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request

**Comment by mattletford** [comment]
@rachelandrew Thanks for the help with adding suggestions to a batch!

**Review by rachelandrew — APPROVED** [review]

**Pull request merged**